internal/app/go-filecoin/retrieval_market_connector/client.go

```diff
@@ -1,27 +1,292 @@
+	aapi ActorAPI,
+	wal WalletAPI,
+	pbapi PaymentBrokerAPI,
+) *RetrievalClientNodeConnector {
+	return &RetrievalClientNodeConnector{
+		bs:         bs,
+		cs:         cs,
+		mw:         mw,
+		outbox:     ob,
+		ps:         ps,
+		sm:         sm,
+		signer:     signer,
+		actAPI:     aapi,
+		wal:        wal,
+		pbAPI:      pbapi,
+		laneStore:  make(map[address.Address]laneEntries),
+	}
+}
+
+// GetOrCreatePaymentChannel gets or creates a payment channel and posts to chain
+// Assumes GetOrCreatePaymentChannel is called before AllocateLane
+// Blocks until message is mined?
+func (r *RetrievalClientNodeConnector) GetOrCreatePaymentChannel(ctx context.Context, clientWallet address.Address, minerWallet address.Address, clientFundsAvailable tokenamount.TokenAmount) (address.Address, error) {
+
+	if clientWallet == address.Undef || minerWallet == address.Undef {
+		return address.Undef, errors.New("empty address")
+	}
+
+	chid, err := r.pbAPI.GetPaymentChannelID(ctx, clientWallet, minerWallet)
+	if err != nil {
+		return address.Undef, err
+	}
+
+	if chid == address.Undef {
+		// create the payment channel
+		bal, err := r.wal.GetBalance(ctx, clientWallet)
+		if err != nil {
+			return address.Undef, err
+		}
+
+		filAmt := types.NewAttoFIL(clientFundsAvailable.Int)
+		if bal.LessThan(filAmt) {
+			return address.Undef, errors.New("not enough funds in wallet")
+		}
+
+		height, err := r.getBlockHeight()
+		if err != nil {
+			return address.Undef, err
+		}
+		validAt := height + 1 // valid almost immediately since a retrieval could theoretically happen in 1 block
+
+		to, err := GoAddrFromFcAddr(fcaddr.LegacyPaymentBrokerAddress)
+		if err != nil {
+			return address.Undef, err
+		}
+
+		fcMiner, err  := FcAddrFromGoAddr(minerWallet)
+		msgCid, _, err := r.outbox.Send(ctx,
+			clientWallet,                           // from
+			to,                                     // to
+			types.ZeroAttoFIL,                      // value
+			types.NewAttoFILFromFIL(1),             // gasPrice
+			types.NewGasUnits(10),                  // gasLimit
+			true,                                   // broadcast to network
+			paymentbroker.CreateChannel,            // command
+			fcMiner, validAt, // params: payment address, valid block height
+		)
+		r.mw.Wait(ctx, msgCid, r.handleMessageReceived)
+
+		return address.Undef, err
+	}
+
+	// Not a real actor, just plays one on PaymentChannels.
+	return chid, nil
+}
+
+// AllocateLane creates a new lane for this paymentChannel with 0 FIL in the lane
+// Assumes AllocateLane is called after GetOrCreatePaymentChannel
 func (r *RetrievalClientNodeConnector) AllocateLane(paymentChannel address.Address) (uint64, error) {
-	panic("TODO: go-fil-markets integration")
+	payer, err := r.getPayerFromChannelID(paymentChannel)
+	if err != nil {
+		return 0, err
+	}
+
+	if r.laneStore[payer] == nil {
+		r.laneStore[payer] = []types.AttoFIL{}
+	}
+	r.laneStore[payer] = append(r.laneStore[payer], types.NewAttoFILFromFIL(0))
```

Comment:
This won't survive a restart. Is this temporary? What is the plan?

Reply:
Any of the stuff stored in memory is temporary. @hannahhoward said she wasn't sure if spec-actors had lanes (but they do) so this will go away.  Otherwise it would have been converted to save to a datastore.